api/rest/coordinator.go

```diff
@@ -0,0 +1,278 @@
+package rest
+
+import (
+	"archive/zip"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+const numberOfWorkers = 10
+const contextDoneErrMsg = "bundle creation context finished before bundle creation finished"
+
+// BundleStatus tracks the status of local bundle creation requests
+type bundleStatus struct {
+	id   string
+	node node
+	done bool
+	err  error
+}
+
+// golangcli-lint marks this as dead code because nothing uses the interface
+// yet. this will be removed with the bundle handler PR
+//nolint:deadcode
+// coordinator is an interface to coordinate the creation of diagnostics bundles
+// across a cluster of nodes
+type coordinator interface {
+	// CreateBundle starts the bundle creation process. Status updates be monitored
+	// on the returned channel.
+	CreateBundle(ctx context.Context, id string, nodes []node) <-chan bundleStatus
+	// CollectBundle waits until all the nodes' bundles have finished, downloads,
+	// and merges them. The resulting bundle zip file path is returned.
+	CollectBundle(ctx context.Context, bundleID string, numBundles int, statuses <-chan bundleStatus) (string, error)
+}
+
+// ParallelCoordinator implements Coordinator interface to coordinate bundle
+// creation across a cluster, parallelized across multiple goroutines.
+type parallelCoordinator struct {
+	client Client
+
+	// statusCheckInterval defines how often the status of the local bundles will
+	// be checked
+	statusCheckInterval time.Duration
+	workDir             string
+
+	quit chan struct{}
+}
+
+// job is a function that will be called by the worker function. The output will be added to results channel
+type job func(context.Context) bundleStatus
+
+// worker is a function that will run incoming jobs from jobs channel and put jobs output to results chan
+func worker(ctx context.Context, jobs <-chan job, results chan<- bundleStatus, quit <-chan struct{}) {
+	for {
+		select {
+		case <-quit:
```

Comment:
In previous commit you remove context, now it's back. Why just not rely on contex timeout instead of adding new channel?

Reply:
I think I was thinking about how this would work incorrectly, doing a context done check there should be fine since the waiting `CollectBundle` loop would still run and create what it can of the bundle.

Reply:
Ok never mind, that causes a deadlock. I think because exiting the worker function will leave behind goroutines that try to push to the `jobs` channel but nothing is reading from it. This is showing up pretty consistently in the tests because jobs isn't buffered right now.